lib/AST/Module.cpp

```diff
@@ -2208,8 +2209,14 @@ SPIGroupsRequest::evaluate(Evaluator &evaluator, const Decl *decl) const {
 }
 
 bool SourceFile::shouldCrossImport() const {
-  return Kind != SourceFileKind::SIL && Kind != SourceFileKind::Interface &&
-         getASTContext().LangOpts.EnableCrossImportOverlays;
+  if (Kind == SourceFileKind::SIL || Kind == SourceFileKind::Interface)
+    return false;
+  // Cross import when experimental differentiable programming is enabled. This
+  // is to enable tgmath derivatives defined in the platform differentiation
+  // overlay.
+  if (isDifferentiableProgrammingEnabled(*this))
```

Comment:
`shouldCrossImport()` gets called during import resolution, and `isDifferentiableProgrammingEnabled()` sometimes calls `namelookup::getAllImports()` to check whether `_Differentiation` is imported, which checks a list that I *believe* is populated by import resolution.

Bottom line: Are you sure that this will work as intended when someone uses `import _Differentiation` rather than the command-line flag? Because I'm definitely not sure it will, and I'm not sure it will continue to work even if it does at the moment. You should at least add a test case.

Reply:
Thanks for the comment. This is a draft PR to get this started. Since I haven't been able to make it even build, I'll need to resolve the high level issues before I can add a test.